overlord/ifacestate/helpers.go

```diff
@@ -540,3 +540,167 @@ func resolveSnapIDToName(st *state.State, snapID string) (name string, err error
+	return false
+}
+
+// Remapping functions for incoming transfers ({state,wire}=>memory)
+
+// RemapIncomingConnRef potentially re-maps connection reference from an API request or being loaded from state.
+//
+// The operation done by remapIncomingConnRef must be symmetric with remapIncomingConnRef.
+// In practice the pair of functions are used to make "snapd" snap the host of implicit
+// interfaces and connections without altering the state in a backwards incompatible way.
+//
+// Data coming from the state and from API requests is changed so that slots on "core"
+// become slots on "snapd" (but only when "snapd" snap itself is being used). When
+// data is about to hit the state again it is re-mapped back.
+func RemapIncomingConnRef(cref *interfaces.ConnRef) (changed bool) {
+	changedPlug := mapper.RemapIncomingPlugRef(&cref.PlugRef)
+	changedSlot := mapper.RemapIncomingSlotRef(&cref.SlotRef)
+	return changedPlug || changedSlot
+}
+
+// RemapIncomingPlugRef potentially re-maps the snap name of a plug reference.
+func RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return mapper.RemapIncomingPlugRef(plugRef)
+}
+
+// RemapIncomingSlotRef potentially re-maps the snap name of a slot reference.
+func RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return mapper.RemapIncomingSlotRef(slotRef)
+}
+
+// Remapping functions for outgoing transfers (memory=>{state,wire})
+
+// RemapOutgoingConnRef potentially re-maps connection reference before being saved to store or returned in an API response.
+//
+// This function is symmetric with RemapIncomingConnRef.
+func RemapOutgoingConnRef(cref *interfaces.ConnRef) (changed bool) {
+	changedPlug := mapper.RemapOutgoingPlugRef(&cref.PlugRef)
+	changedSlot := mapper.RemapOutgoingSlotRef(&cref.SlotRef)
+	return changedPlug || changedSlot
+}
+
+// RemapOutgoingPlugRef potentially re-maps the snap name of a plug reference.
+func RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return mapper.RemapOutgoingPlugRef(plugRef)
+}
+
+// RemapOutgoingSlotRef potentially re-maps the snap name of a slot reference.
+func RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return mapper.RemapOutgoingSlotRef(slotRef)
+}
+
+// RemapOutgoingPlugInfo potentially re-maps the snap name of a plug information.
+//
+// The original object is never altered in place. Instead a new plug info
+// object is returned. The object contains a shallow copy of the original, plus
+// another shallow copy of the associated snap info. The snap info is
+// re-mapped.
+func RemapOutgoingPlugInfo(plugInfo *snap.PlugInfo) *snap.PlugInfo {
+	plugRef := interfaces.PlugRef{Snap: plugInfo.Snap.SnapName(), Name: plugInfo.Name}
+	if changed := mapper.RemapOutgoingPlugRef(&plugRef); changed {
+		remappedSnapInfo := *plugInfo.Snap
+		remappedPlugInfo := *plugInfo
+		remappedPlugInfo.Snap = &remappedSnapInfo
+		remappedSnapInfo.SideInfo.RealName = plugRef.Snap
```

Comment:
mutating in place SnapInfo feels wrong,   snapd and core have potentially very different snap.Infos

Reply:
Those are sadly needed because infos participate in the response API of the interfaces today. See here for how those are used: https://github.com/snapcore/snapd/pull/5491/commits/13903fc018fbbefe4600474f91f7c804dd228c3f#diff-604eb96e84eb2976171c1cedca2bf095R1763

Reply:
This is no longer needed now that daemon/api.go was refactored in master.